### ISSUE TITLE
Support controlling thread affinity

### ISSUE THREAD

**Issue by tmds** [dotnet/corefx#18924]
**Purpose**

Allow control of thread affinity.

**Proposed API**

```C#
class Thread
{
    ProcessorSet Affinity { get; set; }
}
```
The `ProcessorSet` represents a a set of logical processors.
The underlying APIs for Windows/Posix use bit masks for this.

Windows: https://msdn.microsoft.com/en-us/library/windows/desktop/ms686247(v=vs.85).aspx
Posix: http://man7.org/linux/man-pages/man3/pthread_setaffinity_np.3.html

On Windows, systems with more than 64 processors are split into groups. Threads cannot use processors which belong to multiple groups. `ProcessorSet` can store the group info.

```C#
struct ProcessorSet
{
   int Group { get; set; }                  // get/set group
   bool this[int processor] { get; set; }   // get/set/clear processor

   ProcessorSet(minCapacity);               // actual capacity may be higher
   int Capacity { get; }                    // indexer is valid from [0-Capacity[

   // string representation, see below
   string ToString();
   static ProcessorSet Parse(string);
   static bool TryParse(string, out ProcessorSet);

   static ProcessorSet ForProcessor(int group, int processor); // set with single processor
   static ProcessorSet ForProcessor(int processor);            // set with single processor
}
```

The string representation can be a comma separated list of processors and ranges, prefixed with an optional processor group. E.g: `1`, `1-3,5`, `1:8-10`.

**Example**

```
var thread = new Thread(threadStart);
thread.Affinity = ProcessorSet.Parse("0-3");
thread.Start();
```

**Remarks**

This proposal does not consider retrieving information from the OS like the number of groups and the number of processors per group. The total number of processors can be read from `Environment.ProcessorCount`.

Posix threads inherit the calling thread affinity. It may be preferable to allow new Threads to run on all processors to avoid surprises. Threads that want to inherit the calling thread priority can explicitly do so:

```c#
newThread.Affinity = Thread.CurrentThread.Affinity;
```

**Comment by svick** [comment]
There is already `ProcessThread.ProcessorAffinity`. Do I get it right that this proposal is better because of the following reasons?

* `ProcessThread.ProcessorAffinity` throws `PlatformNotSupportedException` on Unix.
* Mapping between `Thread` and `ProcessThread` is difficult.
* `ProcessThread.ProcessorAffinity` does not support groups.

**Comment by tmds** [comment]
@svick Yes, x-plat and more than 64 processors. Limited to the current executing process (compared to ProcessThread).

**Comment by damageboy** [comment]
There's a definite need for having support for affinities > 64 processors...
The undelying Windows/Posix APIs do expose these abilities...

I'd love to have that...
![image](https://user-images.githubusercontent.com/125730/32542967-f4caff84-c47c-11e7-9a02-7c7fda4542d9.png)

**Comment by tmds** [comment]
Closing due to limited interest.

**Comment by tannergooding** [comment]
I think this is something that would be very useful for certain high-performance apps.

**Comment by redwyre** [comment]
AFAIK, there is currently no way to handle more than 64 core using .net even on windows. In fact without access to processor groups the process will end up with access to _less_ cores, We sorely need something like this. AMD is currently pushing core count and it wont be long before 64 core is something we need to handle.
